### ISSUE TITLE
example

### ISSUE THREAD

**Issue by gedw99** [choria-io/go-backplane#10]
I just found this project and have been playing aroudn with it.  It really adds allot of the stuff you need in ops and i like its approach. no more ssh bastions and guesswork etc :)

Is it possible to make an example to show off Choria for a golang apps developer ?

Maybe a docker for the core choria server with the certs setup and then a microservice using the backplane.

**Comment by ripienaar** [comment]
Thanks @gedw99 yeah an example will be published, as you can tell this project started like yesterday so its early days.

I'll probably bake the example into the vagrant demo or something

**Comment by gedw99** [comment]
Sounds great....

Looking forward to trying this out.

**Comment by ripienaar** [comment]
I added a working example in `example` some more things to do to make it usable in the vagrant demo but you can get the idea there

**Comment by gedw99** [comment]
The del is basically a manifest describing what the thing does and what it needs.

I saw that it's codegenerated in the other code somewhere.

Sorry to bug you but it's an usual framework and just trying to get a cognitive reference frame of the how and why..

**Comment by ripienaar** [comment]
Yeah leave this one open I will still work on it

Do you know web services? You would have a JSON Schema or WSDL that describe the service. These DDL and JSON is the same except one configured the Ruby client (DDL) and other the Go one

If you do not know mcollective ... my docs will suck for you. We are mid transition between projects and for new people it’s a bit rough going. 

Start here https://choria.io/docs/concepts/ today the only viable deployment method is with Puppet.

**Comment by gedw99** [comment]
Ok leaving open

It's considerate of you in that yeah I don't know mcollective - it's a terrible rap band name ?  So I gotta read up on the concepts you linked.

I do know SOAP, WSDL, and even a little DISCO from the C# spoonfeeding days. Ironically have to do some WSDL parsing in golang to interface with a Microsoft server at work. Sigh .

**Comment by gedw99** [comment]
Dooh. Reopening ..

**Comment by ripienaar** [comment]
MCollective is a old project I made and sold - ruby framework for writing APIs accessible from the CLI and Ruby API

It’s owners abandoned it and so I started a new project to kick things up again. Focusing on performance, scale etc. We can do a RPC request hitting round trip on 60000 machines in about 1 second 

Traditionally very dependent on Puppet but slowly gaining independence. 

So right now getting all the bits working is some Ruby and some GO and some Puppet CA. Easy for Puppet users hard for others. 

Read the concepts and let me know what you think.

**Comment by gedw99** [comment]
i read the docs and everything there is familiar to me and it looks great. 

Is Puppet part of this architecture going forward or is the idea to move away from it.
TO be brutal, I dont need puppet for what i see Choria doing for me.

**Comment by ripienaar** [comment]
Something has to install and co figure things - Puppet. 

Traditionally we used their CA and so forth but around September they will stop shipping mcollective (we use it’s CLI still) and by then I want to be more or less capable of being standalone. 

Today already you can install it without Puppet it’s quite easy but not documented. Very recently I made the CA pluggable but I am yet to do same On the ruby side for the CLI

**Comment by gedw99** [comment]
ok thats hopefully. It should be pretty easy to replace the Certificate Authority with Vault if thats what you mean by Puppet CA.

Have you got a link to the CA code ?

**Comment by ripienaar** [comment]
Backplane, choria, stream replicator and prometheus streams all use https://github.com/choria-io/go-security/

I won’t abstracts to this level in the ruby code - probably just the manual one with enrollment by a Go App that drops certs on disk

The new CA code isn’t published yet but it will live in the network transport and needs to be able to issue like 1/2 million certs fairly rapidly so it’s quite a undertaking. Could be backed by vault indeed but nodes shouldn’t speak vault directly

**Comment by gedw99** [comment]
Ok. SO i think i should wait until the CA is golang based. I am a puppet
newbie so have no idea...
then i can try things out. Is that ok ?

**Comment by ripienaar** [comment]
yup yup, onboarding non puppet people is definitely something of a goal - but I am some ways off alas, will hopefully get to a point of some kubernetes operator or such, in time - people who fund my work have like zero interest in that so its slow going to get there

**Comment by gedw99** [comment]
Understood. I can't find it. Just a lone Deb. So I will wait until things
are a little bit mature

**Comment by ripienaar** [comment]
@gedw99 the entire CLI and API is in ruby and today provided by Puppet along with its CA etc, Choria exists literally (till quite recently) to make the almost unusably complex to install mcollective usable within 10 to 30 minutes for a Puppet User.

So yeah, it's tough going if you're not the target :) Keep an eye on the mailng list, I'll announce stuff as they come - for backplane specifically I expect I'll publish a binary to manage it independent of all the puppet shit, it's a great project to test the waters with

**Comment by ripienaar** [comment]
@gedw99 what OS are you on? I'll - probably tomorrow - give you 2 binaries one of the example app and one of the new CLI that can run against any gnatsd without any puppet or anything

**Comment by ripienaar** [comment]
or maybe I'll just make a docker compose file

**Comment by ripienaar** [comment]
@gedw99 you can now have a go with docker compose https://github.com/choria-io/go-backplane#docker-demo
